### PR TITLE
Use proc macros to generate conversion functions for MultiLocation

### PR DESCRIPTION
**Pull request by KiChjang**
This PR is a refactoring of the large chunk of manual and boilerplate code that we wrote just to convert `MultiLocation` from different kinds of types -- it's now substituted by a proc macro named `impl_conversion_functions_for_multilocatiion_v1!()`.

### CODE REVIEW

File: xcm/procedural/src/multilocation.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+			fn from(x: Junction) -> Self {
+				MultiLocation { parents: 0, interior: Junctions::X1(x) }
+			}
+		}
+
+		impl From<[Junction; 0]> for MultiLocation {
+			fn from(_: [Junction; 0]) -> Self {
+				MultiLocation { parents: 0, interior: Junctions::Here }
+			}
+		}
```
Comment (apopiak) [inline xcm/procedural/src/multilocation.rs:133]: I'm not sure I'm a fan of this.
It allows for `let multi: MultiLocation = [].into();` correct?

File: xcm/procedural/src/multilocation.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+			fn from((parents, interior): (u8, Junctions)) -> Self {
+				MultiLocation { parents, interior }
+			}
+		}
+
+		impl From<()> for MultiLocation {
+			fn from(_: ()) -> Self {
+				MultiLocation { parents: 0, interior: Junctions::Here }
+			}
+		}
```
Comment (apopiak) [inline xcm/procedural/src/multilocation.rs:115]: I'm not sure I'm a fan of this.
It allows for `let multi: MultiLocation = ().into();` correct?

**Review by apopiak — COMMENTED** [review]
Could we also add some examples and/or compile time tests?
like 
`let multi: MultiLocation = [Parachain(1234), GeneralIndex(42)].into();`

File: xcm/procedural/src/multilocation.rs (hunk 1 above)
Comment (KiChjang) [reply xcm/procedural/src/multilocation.rs:133]: Yeah it does, we already allow this before my changes to the current code, and I'm just keeping it here to not break the public API.

File: xcm/src/v1/multilocation.rs [hunk 3]
```diff
@@ -322,6 +323,7 @@ impl From<Parent> for MultiLocation {
 }
```
Comment (apopiak) [inline xcm/src/v1/multilocation.rs:323]: :+1:

**Review by apopiak — APPROVED** [review]
Nice. +1 for the tests :-)

File: xcm/procedural/src/v0/multilocation.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	Ok(quote! {
+		#from_tuples
+		#from_v1
+	})
+}
+
+fn generate_conversion_from_tuples() -> TokenStream {
+	let mut from_tuples = TokenStream::new();
+
+	for num_junctions in 0..8usize {
```
Comment (kianenigma) [inline xcm/procedural/src/v0/multilocation.rs:37]: could this 8 _not_ be hardcoded?

File: xcm/procedural/src/v0/multilocation.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+			impl From<[Junction; #array_size]> for MultiLocation {
+				fn from(j: [Junction; #array_size]) -> Self {
+					let [#(#idents),*] = j;
+					MultiLocation::#variant( #(#idents),* )
+				}
+			}
+		};
+
+		from_tuples.extend(from_tuple);
+	}
```
Comment (kianenigma) [inline xcm/procedural/src/v0/multilocation.rs:59]:
```suggestion
	from_tuples = (0..8usize).map(|num_junctions| {
		let junctions = (0..=num_junctions).map(|_| format_ident!("Junction")).collect::<Vec<_>>();
		let idents = (0..=num_junctions).map(|i| format_ident!("j{}", i)).collect::<Vec<_>>();
		let variant = &format_ident!("X{}", num_junctions + 1);
		let array_size = num_junctions + 1;

		quote! {
			impl From<( #(#junctions,)* )> for MultiLocation {
				fn from( ( #(#idents,)* ): ( #(#junctions,)* ) ) -> Self {
					MultiLocation::#variant( #(#idents),* )
				}
			}

			impl From<[Junction; #array_size]> for MultiLocation {
				fn from(j: [Junction; #array_size]) -> Self {
					let [#(#idents),*] = j;
					MultiLocation::#variant( #(#idents),* )
				}
			}
		}
	}).collect::<TokenStream2>();
```

File: xcm/procedural/src/v0/multilocation.rs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+				MultiLocation::Null
+			}
+		}
+
+		#from_tuples
+	}
+}
+
+fn generate_conversion_from_v1() -> TokenStream {
+	let mut match_variants = TokenStream::new();
```
Comment (kianenigma) [inline xcm/procedural/src/v0/multilocation.rs:85]: similar suggestion -- you can directly build this variable from an iterator that maps to `quote` and collecting it in `TokenSteam`

**Review by kianenigma — APPROVED** [review]
Looks correct to me, but best if someone else also takes a look and approve before merge

File: xcm/procedural/src/v0/multilocation.rs (hunk 4 above)
Comment (shawntabrizi) [reply xcm/procedural/src/v0/multilocation.rs:37]: yeah, why not an input into the function?

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**